Pinta.Effects/Effects/FeatherEffect.cs

```diff
@@ -0,0 +1,110 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class FeatherEffect : BaseEffect
+{
+
+	public override string Icon => Pinta.Resources.Icons.EffectsDefault;
+
+	public sealed override bool IsTileable => true;
+
+	public override string Name => Translations.GetString ("Feather");
+
+	public override bool IsConfigurable => true;
+
+	public override string EffectMenuCategory => Translations.GetString ("Stylize");
+
+	public FeatherData Data => (FeatherData) EffectData!;  // NRT - Set in constructor
+
+	private readonly IChromeService chrome;
+
+	public FeatherEffect (IServiceProvider services)
+	{
+		chrome = services.GetService<IChromeService> ();
+		EffectData = new FeatherData ();
+	}
+
+	public override void LaunchConfiguration ()
+		=> chrome.LaunchSimpleEffectDialog (this);
+
+	public override void Render (ImageSurface src, ImageSurface dest, ReadOnlySpan<RectangleI> rois)
+	{
+		int radius = Data.Radius;
+		var src_data = src.GetReadOnlyPixelData ();
+		int src_width = src.Width;
+		int src_height = src.Height;
+		var dst_data = dest.GetPixelData ();
+		int dst_width = dest.Width;
+
+		// reset dest to src
+		// Removing this causes preview to not update to lower radius levels
+		foreach (RectangleI roi in rois) {
+			for (int y = roi.Top; y <= roi.Bottom; ++y) {
+				var src_row = src_data.Slice (y * src_width, src_width);
+				var dst_row = dst_data.Slice (y * src_width, src_width);
+				for (int x = roi.Left; x <= roi.Right; x++) {
+					dst_row[x].Bgra = src_row[x].Bgra;
+				}
+			}
+		}
+
+
+		// Collect a list of pixels that surround the object
+		List<PointI> borderPixels = new List<PointI> ();
+		foreach (RectangleI roi in rois) {
+			for (int y = roi.Top; y <= roi.Bottom; ++y) {
+				for (int x = roi.Left; x <= roi.Right; x++) {
+					PointI potentialBorderPixel = new (x, y);
+					if (src.GetColorBgra (src_data, src_width, potentialBorderPixel).A <= Data.TransparencyThreshold) {
+						for (int sx = x - 1; sx <= x + 1; sx++) {
+							for (int sy = y - 1; sy <= y + 1; sy++) {
+								PointI pixel = new (sx, sy);
+
+								if (sx < 0 || sx >= src_width || sy < 0 || sy >= src_height)
+									continue;
+
+								if (src.GetColorBgra (src_data, src_width, pixel).A != 0)
+									borderPixels.Add (potentialBorderPixel);
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// For each pixel, lower alpha based off distance to border pixel
+		foreach (var borderPixel in borderPixels) {
+			for (int y = borderPixel.Y - radius; y <= borderPixel.Y + radius; ++y) {
+				for (int x = borderPixel.X - radius; x <= borderPixel.X + radius; x++) {
```

Comment:
I think the issue with modifying the image outside the selection is because these coordinates aren't clamped to the regions being rendered (or the image size even?)

Reply:
The issue I am having, is that the border pixels have to affect pixels outside their roi. I can't make the system work backwards (each pixel modifies itself based on nearby border pixels) because of threading (Pixels may not be aware of some border pixels.) I dont think it's possible to see all the rois at once.

That being said, yeah i should be clamping to at least the border size...

Reply:
I think if you define `IsTileable` to be false, then you'll get a single region for the whole image (it's up to you to then do any multithreading internally)
That could be more suitable here if you want to do a first pass to find border pixels and then a second pass to adjust alpha, or something like that?